### DISCUSSION TITLE
Any chance of bin/cue support?

### DISCUSSION THREAD

**Discussion by derkommutator** [ZuluSCSI/ZuluSCSI-firmware#143] [Ideas]
I've got an installation CD I'm hoping to use with ZuluSCSI (specifically a BeOS 5.0.3 CD), which is multi-session and therefore in bin/cue format rather than iso.

I know I can use bchunk to break it apart into 3 isos, but I'm anticipating running into trouble when the installer neither gives an option to change discs nor provides a software eject button.

I don't yet have the ZuluSCSI I've ordered, so I haven't tested this out yet, and may be able to work around the problem somehow, but I figured I'd still start this thread to see if bin/cue is on the roadmap, or if there's any interest in adding it.

**Comment by PetteriAimonen** [comment]
I think the .bin file could work directly, though it depends on what kind of tracks the image uses.

But I agree that for some CDs like ones that mix audio & data tracks there could be need to support data from the cue file also.
Currently the CDROM TOC is hardcoded in https://github.com/ZuluSCSI/ZuluSCSI-firmware/blob/main/lib/SCSI2SD/src/firmware/cdrom.c

**Comment by derkommutator** [comment]
Moi Petteri.  :)  Okay, when I've got the hardware in hand, I'll give it a shot with the .bin file directly and see what happens.  In the case of this disc, it's all data, one boot session followed by an x86 session and a PPC session, so perhaps it will indeed just work.

As you note, for images that mix in audio, this approach certainly doesn't work.  FYI there's a similar project, aimed primarily at classic Macs, but also compatible with other hardware, that does implement bin/cue and is known to work (on Macs at least) with audio data:  https://macsd.com/

I'll update this thread with my findings once I receive the hardware and try the .bin file.

**Reply by aperezbios** [reply]
@derkommutator In the case of data-only .bin files, you should be able to use a utility to re-assemble a multi-track BIN to an ISO file. I've done so successfully in the past.

**Comment by derkommutator** [comment]
Just a quick update.  I finally got my ZuluSCSI (labour strike kept it at the port for a while), and unfortunately I've not yet been able to find out how well it'll handle the .bin file, because so far I've not been successful in getting the BeBox to read even known good .iso files.

A disk image of the old spinning SCSI drive created via ZuluSCSI's initiator mode works fine for booting the old BeOS that was already installed, but I can't boot the CD-ROM's .bin, nor .isos extracted from it with bchunk, and neither can existing BeOS read a plain old known good unrelated .iso.

Any attempt to access the virtual CD-ROM device results in a scsi_accel_rp2040_stopWrite() timeout.  I am running the latest snapshot firmware, so it doesn't lock up the bus anymore, but I still can't read the .iso.

Unfortunately, owing to the rarity of the BeBox hardware, I'm probably the only person who has ever tried a ZuluSCSI on one.  :)  I've run through all the zuluscsi.ini config options, trying out basically everything, with no resolution.

Although it's no longer on-topic for this thread, I've attached the zululog.txt.  I can open an issue if you like, or keep on poking at it myself.  I do realise that this is an incredibly niche use of the ZuluSCSI, and I'm quite possibly the only person on the planet who will ever try.  ;)

[zululog.txt](https://github.com/ZuluSCSI/ZuluSCSI-firmware/files/10954274/zululog.txt)

**Reply by PetteriAimonen** [reply]
I think this may be related to #149, where some progress has been made by setting the block size to 512. So you could try names `CD6_512.iso` and `CD6_1024.iso`. The current default for CD drives is 2048 because that is what Mac expects, but looks like it is not universal.

**Reply by MeTalOneOEight** [reply]
If I read the SCSI2SD CD drive sources correctly, it does only support track 1 thus multi-session CDs only work with the first session.

So maybe more peeking at #149 may suggest using the image with HD name instead of CD could be worth a try.

**Reply by derkommutator** [reply]
Ding ding ding.  I'm sorry I didn't notice #149 before posting, as that indeed has solved the problem for me.  The bchunk-extracted PPC .iso, renamed HDx.img, boots as hoped for, and now I am able to do the BeOS install.  Thanks for the tip!

Of course this renders this whole thread moot, since this method does not work for trying out the whole .bin file.  Oh well.  :)

**Reply by derkommutator** [reply]
Oh, and to follow up on Petteri's suggestion, no, changing the block size sadly had no effect on the CD emulation.

**Reply by MeTalOneOEight** [reply]
Did you try the whole .bin with HD name?

**Reply by derkommutator** [reply]
Yeah.  The BeBox does not detect it as a bootable device.

**Comment by aperezbios** [comment]
@derkommutator as luck would have it, I actually have a (dual 133MHz) BeBox. Thanks for bringing this up. We'll see if we can make any progress with it.

**Reply by derkommutator** [reply]
What a lucky coincidence!  :)  With my problem now solved by simply renaming the .iso to HDx.img, this looks like it's no longer a problem for me, but of course being able to claim full BeBox compatibility for the ZuluSCSI will be worth some nerd cred.  😆

**Reply by aperezbios** [reply]
Thanks, this is really helpful information.

**Comment by MeTalOneOEight** [comment]
There is some GPL licensed lib for accessing cue/bin, currently alpha status.
Don't know if it's usuable and how large the effort would be to include.
https://github.com/libyal/libodraw

Not sure if that would also mean to implement multi track support for the CD sources.

**Comment by PetteriAimonen** [comment]
Basic bin/cue support has now been implemented (#204).
Test binaries are available here, and it will be part of next release:
https://github.com/ZuluSCSI/ZuluSCSI-firmware/releases/tag/latest

Turned out that the actual cue file parsing was the easy part - more work was needed on the SCSI multi-media commands extension implementation. The previous code only implemented access to data tracks, now there is basic support of accessing audio tracks.

Actual audio playback (to external speakers) has not yet been implemented, but there is work going on it (it will need extra hardware also).

But if anyone is interested, it would be great if you can test it with bin/cue images.
To use, just name both files with same base, e.g. `CD3.bin` and `CD3.cue`.
There is more details in the [README](https://github.com/ZuluSCSI/ZuluSCSI-firmware#cd-rom-images-in-bincue-format).

**Comment by aperezbios** [comment]
BIN/CUE support is complete, @derkommutator. Please download the latest release and give it a try, if you haven't already.
